src/ripple/app/main/NodeStoreScheduler.cpp

```diff
@@ -48,16 +46,21 @@ void NodeStoreScheduler::onChildrenStopped ()
 void NodeStoreScheduler::scheduleTask (NodeStore::Task& task)
 {
     ++m_taskCount;
-    m_jobQueue->addJob (
+    if (!m_jobQueue->addJob (
         jtWRITE,
         "NodeObject::store",
-        [this, &task] (Job&) { doTask(task); });
+        [this, &task] (Job&) { doTask(task); }))
+    {
+        // Job not added, presumably because we're shutting down.
+        // Recover by executing the task synchronously.
+        doTask (task);
+    }
 }
 
 void NodeStoreScheduler::doTask (NodeStore::Task& task)
 {
     task.performScheduledTask ();
-    if ((--m_taskCount == 0) && isStopping())
+    if ((--m_taskCount == 0) && isStopping() && areChildrenStopped())
```

Comment:
Reading the [Stoppable documentation](https://github.com/ripple/rippled/blob/develop/src/ripple/core/Stoppable.h#L125-L163), I suspect the `NodeStoreScheduler` was written to just follow the example code there.  I don't think `NodeStoreScheduler` has any extra asynchronous logic outside of the JobQueue tasks, which is why `NodeStoreScheduler::onChildrenStopped` has `assert(m_taskCount == 0)`.  If you imagined `m_taskCount` were counting tasks in some local thread pool (rather than the child `Stoppable` `JobQueue`), then `NodeStoreScheduler::onChildrenStopped` could only call `stopped` if the imagined internal thread pool were empty at that time.  If it were not empty and `doTask` was servicing that thread pool, it would be responsible for calling `stopped()` once the last task finished.   Therefore, I think your fix is a good one, as you should not be calling `stopped` if you don't know your children have stopped.

So TL;DR, I think your fix is a good one to match the general `Stoppable` framework. However, I suspect given that `NodeStoreScheduler` has internal logic to wait on, we could probably rewrite `doTask` as below and leave `NodeStoreScheduler::onChildrenStopped` responsible for signaling `stopped`.
```
void NodeStoreScheduler::doTask (NodeStore::Task& task)
{
    task.performScheduledTask ();
}
```
I'm still not advocating heavily for such a change, just want to put my thoughts down.